File: 프로그래머스/3/43164. 여행경로/여행경로.js
function solution(tickets) {
    tickets.sort();
    const n = tickets.length;
    let q = ['ICN'];
    let used = Array(n).fill(false);
    
    function dfs(cnt, cur) {
        if (cnt === n) return true;
        for (let i = 0; i < n; i++) {
            const [from, to] = tickets[i];
            if (used[i] || from !== cur) continue;
            q.push(to);
            used[i] = true;
            if (dfs(cnt + 1, to)) return true;
            
            q.pop();
            used[i] = false;
        }
        return false;
    }
    dfs(0, 'ICN');
    return q;
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
import sys
sys.setrecursionlimit(10**6)

def solution(tickets):
    n = len(tickets)
    tickets.sort()
    visited = [False] * n
    ans = []
    
    def dfs(route):
        nonlocal ans
        
        if len(route) == n + 1:
            ans = route[:]
            return True
        
        cur = route[-1]
        
        for i in range(n):
            start, end = tickets[i]
            
            if visited[i] or start != cur: continue
            
            visited[i] = True
            route.append(end)
            
            if dfs(route): return True
            
            route.pop()
            visited[i] = False
        
        return False
    
    dfs(['ICN'])
    
    return ans